flocker/node/agents/test/test_blockdevice.py

```diff
@@ -1198,16 +1205,231 @@ class DidNotConverge(Exception):
     """
-    def setUp(self):
-        self.deployer = create_blockdevicedeployer(self)
 
-    def teardown_example(self, token):
+    def __init__(self, dataset_id, node_uuid, mountroot, blockdevice_manager):
+        self._known_mountpoints = set()
+        self._device_path = None
+        self._dataset_id = dataset_id
+        self._node_uuid = node_uuid
+        self._testing_mountroot = mountroot
+        self._blockdevice_manager = blockdevice_manager
+
+    def _has_file_with_content(self, filename, content):
+        """
+        Returns whether the blockdevice at self._device_path has a file with
+        the given name.
+
+        :param unicode filename: The name of the file to check for.
+        :param unicode content: The expected content of the file.
+
+        :returns: Whether the blockdevice that backs ``self._dataset_id`` has a
+            file with the given name that contains the given content.
+        """
+        test_mountpoint = self._testing_mountroot.child(str(uuid4()))
+        test_mountpoint.makedirs()
+
+        # If the device is already mounted, read the file from the existing
+        # mountpoint.
+        for mount_info in self._blockdevice_manager.get_mounts():
+            if mount_info.blockdevice == self._device_path:
+                file_to_test = mount_info.mountpoint.child(filename)
+                return (file_to_test.exists() and
+                        file_to_test.getContent() == content)
+
+        # Otherwise mount the blockdevice and read the file from the new
+        # mountpoint.
+        try:
+            self._blockdevice_manager.mount(self._device_path, test_mountpoint)
+            file_to_test = test_mountpoint.child(filename)
+            result = (file_to_test.exists() and
+                      file_to_test.getContent() == content)
+            self._blockdevice_manager.unmount(self._device_path)
+        except MountError:
+            # We might not be able to mount if we do not have a filesystem on
+            # the blockdevice.
+            return False
+        finally:
+            test_mountpoint.remove()
+        return result
+
+    def invariant(self, case, current_cluster_state):
+        """
+        An invariant that should always hold. Specifically, after the path
+        to a dataset is known by the cluster then writes to that path
+        should either succeed and be persisted to the underlying block
+        device, or the write should fail.
+
+        :param case: A ``TestCase`` to use for making assertions.
+        :param current_cluster_state: A callable that returns the current state
+            of the cluster.
+        """
+        dataset = unicode(self._dataset_id)
+
+        # First poll the cluster state to see if there is a new mountpoint that
+        # we should try to use to attempt to write to the dataset.
+        [node] = list(n for n in current_cluster_state().nodes
+                      if n.uuid == self._node_uuid)
```

Comment:
Following up on the discussion [here](https://github.com/ClusterHQ/flocker/pull/2339#discussion_r47719027), it occurs to me that the state you are inspecting is *still* internal state. If you want the externally visible state, you'd need to look at the result of [this](https://github.com/ClusterHQ/flocker/blob/master/flocker/control/httpapi.py#L466). I don't think it is worth it to reach that far out, so please use the locally discovered state rather than the cluster state. Part of the reason is that I hope the later will change significantly in the near future, and don't want to have to have another place that references it.